### PR TITLE
[irods/irods#2840][irods/irods#4305] Expand ils help text (master)

### PR DESCRIPTION
**Pull request by alanking**
This is just a start. Please leave suggestions on what should be included here.

In the original issue, the author wanted definitions for the generic data type, replica numbers, etc. but I felt these may be more appropriate in other documentation (which I can make in irods/irods_docs). I will include them if others feel differently.

### CODE REVIEW

File: src/ils.cpp [hunk 1]
```diff
@@ -101,6 +101,66 @@ usage() {
         "Usage: ils [-ArlLv] dataObj|collection ... ",
         "Usage: ils --bundle [-r] dataObj|collection ... ",
         "Display data Objects and collections stored in irods.",
```
Comment (trel) [inline src/ils.cpp:103]: Objects -> objects

Comment (trel) [reply src/ils.cpp:103]: irods -> iRODS

File: src/ils.cpp [hunk 2]
```diff
@@ -101,6 +101,66 @@ usage() {
...
+        " ",
+        "The following is typical output for ils with no arguments:"
+        "    $ ils",
+        "    /tempZone/public:",
+        "      big",
+        "      foo",
+        "The only information displayed are the names of data objects in the current",
+        "working directory.",
```
Comment (trel) [inline src/ils.cpp:111]: directory -> collection?

File: src/ils.cpp [hunk 3]
```diff
@@ -101,6 +101,66 @@ usage() {
...
+        " ",
+        "The long format (given by -l) will display some more information:",
+        "    $ ils -l",
+        "    /tempZone/public:",
+        "      rods              0 demoResc     41943040 2021-04-29.17:57 & big",
+        "      rods              2 comp;arch    40236581 2021-04-19.03:00 X big",
```
Comment (trel) [inline src/ils.cpp:117]: i think bump the filesize to the right a few spaces to align with the 'foos' below.

File: src/ils.cpp [hunk 4]
```diff
@@ -101,6 +101,66 @@ usage() {
...
+        "the owner of the data object and the replica's number, size, last modified",
+        "time, and status.",
+        " ",
+        "The very long format (given by -L) will display even more information:",
+        " ",
+        "    $ ils -L",
+        "    /tempZone/public:",
+        "      rods              0 demoResc     41943040 2021-04-29.17:57 & big",
+        "        sha2:gKNyEYjkAhiwiyZ3a8U72ugeR4T/9x1xRQoZcxnLoRM=    generic    /var/lib/irods/Vault/home/rods/big",
+        "      rods              2 comp;arch    40236581 2021-04-19.03:00 X big",
```
Comment (trel) [inline src/ils.cpp:132]: same, bump the filesize lines to the right to align with the foos

File: src/ils.cpp [hunk 5]
```diff
@@ -101,6 +101,66 @@ usage() {
...
+        "The long format (given by -l) will display some more information:",
+        "    $ ils -l",
+        "    /tempZone/public:",
+        "      rods              0 demoResc     41943040 2021-04-29.17:57 & big",
+        "      rods              2 comp;arch    40236581 2021-04-19.03:00 X big",
+        "      alice             2 repl1;child1          283 2021-04-29.17:54 ? foo",
+        "      alice             3 repl1;child2          283 2021-04-29.17:56 ? foo",
+        "      alice             4 repl1;child3          283 2021-04-29.17:57 ? foo",
+        "This displays replica information and system metadata for the replicas of",
+        "the data objects in the current working directory. This information includes",
```
Comment (trel) [inline src/ils.cpp:122]: directory -> collection

File: src/ils.cpp [hunk 6]
```diff
@@ -101,6 +101,66 @@ usage() {
...
+        "the owner of the data object and the replica's number, size, last modified",
```
Comment (trel) [inline src/ils.cpp:123]: and resource hierarchy

File: src/ils.cpp [hunk 7]
```diff
@@ -101,6 +101,66 @@ usage() {
...
+        "            generic    /var/lib/irods/child1vault/home/rods/foo",
+        "      alice             3 repl1;child2          283 2021-04-29.17:56 ? foo",
+        "            generic    /var/lib/irods/child2vault/home/rods/foo",
+        "      alice             4 repl1;child3          283 2021-04-29.17:57 ? foo",
+        "            generic    /var/lib/irods/child2vault/home/rods/foo",
+        "This displays even more replica information and system metadata such as each ",
+        "replica's checksum, data type, and physical location in the storage resource.",
+        " ",
+        "The different replica statuses are represented by the following characters ",
+        "which can be seen next to the data object name on the right of an ils -l/-L:",
```
Comment (trel) [inline src/ils.cpp:144]: 'seen between the last modified time and the data object name'

File: src/ils.cpp [hunk 8]
```diff
@@ -101,6 +101,66 @@ usage() {
...
+        "   0. X: stale        - The replica is no longer known to be good. Usually",
+        "                        indicates that a more recently written-to replica exists.",
+        "                        This does not necessarily mean that the data is incorrect",
+        "                        but merely that it may not reflect the \"truth\" of this",
+        "                        data object.",
+        "   1. &: good         - The replica is good. Usually the latest replica to be",
+        "                        written. When a replica is good, all bytes are understood",
```
Comment (trel) [inline src/ils.cpp:151]: and system metadata (size and checksum, if present)

**Review by trel — COMMENTED** [review]
looks pretty good... let's look at it again in a couple more days

File: src/ils.cpp (hunk 1 above)
Comment (alanking) [reply src/ils.cpp:103]: Done

File: src/ils.cpp (hunk 2 above)
Comment (alanking) [reply src/ils.cpp:111]: Done

File: src/ils.cpp (hunk 3 above)
Comment (alanking) [reply src/ils.cpp:117]: Got it

File: src/ils.cpp (hunk 4 above)
Comment (alanking) [reply src/ils.cpp:132]: Done

File: src/ils.cpp (hunk 5 above)
Comment (alanking) [reply src/ils.cpp:122]: Done

File: src/ils.cpp (hunk 6 above)
Comment (alanking) [reply src/ils.cpp:123]: Yup

File: src/ils.cpp (hunk 7 above)
Comment (alanking) [reply src/ils.cpp:144]: Done

File: src/ils.cpp (hunk 8 above)
Comment (alanking) [reply src/ils.cpp:151]: Done

File: src/ils.cpp [hunk 9]
```diff
@@ -100,56 +100,57 @@ usage() {
...
         " ",
         "The long format (given by -l) will display some more information:",
         "    $ ils -l",
         "    /tempZone/public:",
-        "      rods              0 demoResc     41943040 2021-04-29.17:57 & big",
-        "      rods              2 comp;arch    40236581 2021-04-19.03:00 X big",
-        "      alice             2 repl1;child1          283 2021-04-29.17:54 ? foo",
-        "      alice             3 repl1;child2          283 2021-04-29.17:56 ? foo",
-        "      alice             4 repl1;child3          283 2021-04-29.17:57 ? foo",
+        "      rods              0 demoResc              41943040 2021-04-29.17:57 & big",
```
Comment (trel) [inline src/ils.cpp:116]: gah, this is not better.    i see now with local testing that they don't actually all line up like this together... they're kind of all over the place.

if we're going to manipulate this... can we right justify the file size column?  i think that's what will be most readable.

Comment (trel) [reply src/ils.cpp:116]: very good - yes please.

File: src/ils.cpp [hunk 10]
```diff
@@ -100,7 +100,68 @@ usage() {
...
+        "last modified time, and status.",
+        " ",
+        "The very long format (given by -L) will display even more information:",
+        " ",
+        "    $ ils -L",
+        "    /tempZone/public:",
+        "      rods              0 demoResc              41943040 2021-04-29.17:57 & big",
+        "        sha2:gKNyEYjkAhiwiyZ3a8U72ugeR4T/9x1xRQoZcxnLoRM=    generic    /var/lib/irods/Vault/home/rods/big",
+        "      rods              2 comp;arch             40236581 2021-04-19.03:00 X big",
+        "        sha2:KNyEYjkhiwyZ3a8U72ugeR4T/9x1xRQoZcxnLoRMigg=    generic    /var/lib/irods/Vault/home/rods/big",
```
Comment (trel) [inline src/ils.cpp:133]: these two physical paths are the same... can you change the arch path to be archVault?

Comment (trel) [reply src/ils.cpp:133]: after that - i think we're good for squashin'

Comment (alanking) [reply src/ils.cpp:133]: Good call. These should probably all be in public, also.

**Comment by alanking** [comment]
Squashed and #'d

**Pull request merged**